src/nvim/option.c

```diff
@@ -3731,69 +3598,148 @@ int get_option_value_strict(char *name, int64_t *numval, char **stringval, int o
 ///
-/// The exact semantics of this depend on the option.
-static OptVal clear_optval(const char *name, uint32_t flags, void *varp, buf_T *buf, win_T *win)
+/// @param       opt_idx    Option index. Must be >=0.
+/// @param[in]   varp       Option variable pointer, cannot be NULL.
+/// @param       value      New option value.
+/// @param       opt_flags  Option flags.
+/// @param       new_value  Whether value was replaced completely.
+/// @param[out]  errbuf     Buffer for error message.
+/// @param       errbuflen  Length of error buffer.
+///
+/// @return  Error message. NULL if there are no errors.
+static const char *set_option(const int opt_idx, void *varp, OptVal value, int opt_flags,
+                              const bool new_value, char *errbuf, size_t errbuflen)
 {
-  OptVal v = NIL_OPTVAL;
+  assert(opt_idx >= 0 && varp != NULL);
 
-  // Change the type of the OptVal to the type used by the option so that it can be cleared.
-  // TODO(famiu): Clean up all of this after set_(num|bool|string)_option() is unified.
+  const char *errmsg = NULL;
+  bool value_checked = false;
+  vimoption_T *opt = &options[opt_idx];
 
-  if (flags & P_BOOL) {
-    v.type = kOptValTypeBoolean;
-    if ((int *)varp == &buf->b_p_ar) {
-      // TODO(lewis6991): replace this with a more general condition that
-      // indicates we are setting the local value of a global-local option
-      v.data.boolean = kNone;
-    } else {
-      v = get_option_value(name, NULL, OPT_GLOBAL, NULL);
-    }
-  } else if (flags & P_NUM) {
-    v.type = kOptValTypeNumber;
-    if ((OptInt *)varp == &curbuf->b_p_ul) {
-      // The one true special case
-      v.data.number = NO_LOCAL_UNDOLEVEL;
-    } else if ((OptInt *)varp == &win->w_p_so || (OptInt *)varp == &win->w_p_siso) {
-      // TODO(lewis6991): replace this with a more general condition that
-      // indicates we are setting the local value of a global-local option
-      v.data.number = -1;
-    } else {
-      v = get_option_value(name, NULL, OPT_GLOBAL, NULL);
+  // TODO(famiu): Unify set_string_option with set_option.
+  if (value.type == kOptValTypeString) {
+    errmsg = set_string_option(opt_idx, varp, value.data.string.data, opt_flags, true,
+                               &value_checked, errbuf, errbuflen);
+    goto end;
+  }
+
+  // Disallow changing some options from secure mode.
+  if ((secure || sandbox != 0) && (options[opt_idx].flags & P_SECURE)) {
+    return e_secure;
+  }
+
+  OptVal old_value = optval_from_varp(value.type, varp);
+  OptVal old_global_value = NIL_OPTVAL;
+  OptVal old_local_value = NIL_OPTVAL;
+
+  // Save the local and global values before changing anything. This is needed as for a global-only
+  // option setting the "local value" in fact sets the global value (since there is only one value).
+  //
+  // TODO(famiu): This needs to be changed to use the current type of the old value instead of
+  // value.type, when multi-type options are odded.
```

Comment:
typo?

also is this TODO tracked somewhere, like in a tracking issue for the planned options-related work?

Reply:
Yup, that is a typo. And yes, it is not tracked. Should I make a tracking issue detailing the things I wish to work on?

Reply:
> Should I make a tracking issue detailing the things I wish to work on?

That could be helpful, unless you have them in your head and will address them in a nearby PR :)

Reply:
I think it's best if I make an issue so I can have a concrete roadmap of things to work on, and others can know what I plan to do with this as well. Will do that in a bit